[Fleet] Change agent policies in edit package policy page

juliaElastic opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/184394

Added agent policy selection to Edit integration policy page.


There is a lot of duplication between Create and Edit integration policy pages, I'll see if I can refactor to extract the common logic: steps components and managing its state.
I extracted the steps to a hook, it would be a bigger refactor to use this in Create package policy page, so I might create a follow up issue for that.

## To verify
- enable the `enableReusableIntegrationPolicies` experimental feature in `kibana.dev.yml`
- Create a few agent policies
- Add an integration 
- Go to Edit integration, and modify the linked agent policies
- Verify that the existing agent policies are populated correctly in the Existing hosts combo box
- Verify that the modified agent policy list is reflected in the `Preview API Request`, `policy_ids` list.
- Verify that when submitting the form, the package policy linkages are updated to the selected ones (add/remove agent policies)
- The agent count should update below the combo / in the submit modal window
- It's not allowed to submit the form after removing all agent policies
- If a new agent policy is selected, it will be created first and then assigned to the integration policy

<img width="995" alt="image" src="https://github.com/elastic/kibana/assets/90178898/0a7163c6-154e-49b1-b73c-19ed024f6dc3">
<img width="993" alt="image" src="https://github.com/elastic/kibana/assets/90178898/ad470a27-90fa-40f5-b394-a93a08c95e06">
<img width="535" alt="image" src="https://github.com/elastic/kibana/assets/90178898/3b0ddc29-abf8-4e0d-8beb-300634c245b3">
<img width="1758" alt="image" src="https://github.com/elastic/kibana/assets/90178898/e8b976fe-3e53-439c-9b23-803deaf3e0aa">

### Create agent policy
<img width="1737" alt="image" src="https://github.com/elastic/kibana/assets/90178898/6f2a7f65-981a-487d-87c4-2dbb7ecd1835">
Preview API request contains the POST agent policy request
<img width="896" alt="image" src="https://github.com/elastic/kibana/assets/90178898/109140ab-13f2-42c9-9bbc-fb64859c4f62">
After submit, the updated integration policy is assigned to the new agent policy too
<img width="2552" alt="image" src="https://github.com/elastic/kibana/assets/90178898/4027b47b-8d20-4153-b7ec-ed3500f08c9a">


## Open questions

- Currently the namespace placeholder of the package policy is set to show the namespace of the first selected agent policy (if not set by the package policy). I have to check what happens on the backend, if the inherited namespace is changed if the agent policies change. The behaviour should be consistent in the backend and UI.
Currently on the Agent policy details UI, the same integration policy might show different inherited namespace if its shared by multiple agent policies with different namespace.

<img width="1498" alt="image" src="https://github.com/elastic/kibana/assets/90178898/567800a8-2dcb-4b18-af89-f6e902889092">
<img width="1326" alt="image" src="https://github.com/elastic/kibana/assets/90178898/b59d131e-314c-4d5a-81e3-ab8fe0fa6e1f">
<img width="1318" alt="image" src="https://github.com/elastic/kibana/assets/90178898/69b54a63-f7c1-4f0f-8041-74b1774f1e9e">

- When the Edit integration was started from the Agent policy details UI, the navigation goes back to the same agent policy after submitting the page. Is this okay? Might be somewhat unexpected if creating a new agent policy, though it is getting complex to decide where to navigate in case of multiple agent policies.


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

juliaElastic commented:
/ci

juliaElastic commented:
/ci

juliaElastic commented:
/ci

juliaElastic commented:
/ci

juliaElastic commented:
/ci

juliaElastic commented:
/ci

criamico commented:
> I extracted the steps to a hook, it would be a bigger refactor to use this in Create package policy page, so I might create a follow up issue for that.

Thanks for this, it's already a huge step! ❤️

criamico commented:
> Currently on the Agent details UI, the same integration policy might show different inherited policy if its shared by multiple agent policies with different namespace.
So I'm testing locally, so I created an integration policy with two different agent policies. Then I changed only one of the agent policies to a new namespace, I went back to edit the integration policy and noticed that the namespace is updated as well, i.e.

- Agent policy 1 -> namespace 'default'
- Agent policy 2 -> namespace was 'default', updated to 'testspace'
- integration policy -> namespace "automagically" updated to 'testspace'

I don't know if this how it should behave, product can give more context for sure.
However I suggest to file a ticket to tweak the namespace flow on its own and to update the copy under the namespace input so users will know what to expect.

criamico commented:
I also have an observation regarding the "new hosts" flow. This was created with the expectation that you could only associate one integration policy to with a single agent policy, so it uses two tabs. However now that the user can add multiple policies, it's not very clear how it will behave.

For instance, I edited an integration that already had two policies and created a new one from the "new hosts" tab. The UI doesn't really clarify what the user should expect: the newly created policy will override the other two or will be added to them? It's clear that adds the new policy to the other two only `after` submitting. Should we try to clarify this behavior in the UI?

jen-huang commented:
> When the Edit integration was started from the Agent policy details UI, the navigation goes back to the same agent policy after submitting the page. Is this okay? Might be somewhat unexpected if creating a new agent policy, though it is getting complex to decide where to navigate in case of multiple agent policies.

IMO this is ok, the design spec actually specifically calls out this flow as going back to the original agent policy that the user initiated from.

nimarezainia commented:
> > When the Edit integration was started from the Agent policy details UI, the navigation goes back to the same agent policy after submitting the page. Is this okay? Might be somewhat unexpected if creating a new agent policy, though it is getting complex to decide where to navigate in case of multiple agent policies.
> 
> IMO this is ok, the design spec actually specifically calls out this flow as going back to the original agent policy that the user initiated from.

I also think this is ok.

nimarezainia commented:
> > For instance, I edited an integration that already had two policies and created a new one from the "new hosts" tab. The UI doesn't really clarify what the user should expect: the newly created policy will override the other two or will be added to them? It's clear that adds the new policy to the other two only after submitting. Should we try to clarify this behavior in the UI?
> 
> Yes, I've been thinking about this too, currently the new agent policy is added to the existing policies, and any changes to the `Existing hosts` tab will be persisted too. I think it would be great to summarize the list of agent policies that will be persisted in the submit confirm modal. I'll ask for confirmation from Nima/Simona.

@juliaElastic @criamico (thanks for all the work here) when you are in the `Existing Hosts` tab, multiple agent policies can be added. What is the "submit confirm" modal referring to here? The user sees the list of agent policies already so not sure if a confirmation is required. Unless I am missing the point of where this modal will reside.

juliaElastic commented:
> > Currently on the Agent details UI, the same integration policy might show different inherited namespace if its shared by multiple agent policies with different namespace.
> 
> I'm testing locally, so I created an integration policy with two different agent policies. Then I changed only one of the agent policies to a new namespace, I went back to edit the integration policy and noticed that the namespace is updated as well, i.e.
> 
> * Agent policy 1 -> namespace 'default'
> * Agent policy 2 -> namespace was 'default', updated to 'testspace'
> * integration policy -> namespace "automagically" updated to 'testspace'
> 
> I don't know if this how it should behave, product can give more context for sure. However I suggest to file a ticket to tweak the namespace flow on its own and to update the copy under the namespace input so users will know what to expect.

I checked what happens on the backend when an integration policy is used in multiple agent policies with different namespace. It turns out the agents correctly send data using the namespace of the agent policy they are enrolled to, so I think it works correctly.
The only misleading bit is on the Edit integration UI is where we show the first agent policy's namespace as a placeholder if the integration policy doesn't define its own namespace. Instead, we may want to show something else e.g. `inherited` with a tooltip saying the integration policy doesn't define its own namespace, and the agent policy's will be used when ingesting data.

<img width="1546" alt="image" src="https://github.com/elastic/kibana/assets/90178898/152547d6-fef6-4b6c-a704-372938475805">
<img width="1243" alt="image" src="https://github.com/elastic/kibana/assets/90178898/8e145a6a-90d6-4808-91e3-99dc0c35b557">

juliaElastic commented on x-pack/plugins/fleet/public/applications/fleet/sections/agent_policy/create_package_policy_page/components/steps/step_select_agent_policy.tsx:
this change fixes a bug where a validation error was shown on Edit package policy page when the package is limited (e.g. endpoint) and the existing agent policies were selected (right when opening the form without any change in agent policies)
<img width="993" alt="image" src="https://github.com/elastic/kibana/assets/90178898/658abb8a-dce0-4386-9706-bd00e9505275">

juliaElastic replied:
There is still a bug where `Save integration` button is enabled on Edit integration page without making any changes. This is because the `packagePolicy.policy_ids` are being changed multiple times while loading the existing policies in the combo box.

criamico replied:
I've taken a look at this and I don't think it is strictly a "bug": if we disable the save integration button on load, it will prevent the user from creating a new policy. 

- Editing the integration with creation of a new agent policy - button is enabled:

https://github.com/elastic/kibana/assets/16084106/5de05954-691a-4dca-ad78-62841da6fe16

- Trying the same with button disabled at loading 

https://github.com/elastic/kibana/assets/16084106/24ee20ad-914e-4752-a165-1f52ecc35e28

In this case it would be possible to add an existing agent policy from the "existing hosts" tab, but it's not possible to assign the integration to a fresh policy. So I think it makes sense to keep the "save integration" button enabled at loading to avoid this issue.

kpollich replied:
Thanks for taking a look at this @criamico - do you think it's worth moving this caveat to a follow-up issue and discussing further?

criamico replied:
I think it would be better, as the "create new policy" flow doesn't look very intuitive to me as it is. That selector was created having in mind the association `1 integration: 1 agent policy` but now that there can be multiple policies it's not very suited for it. I think it would make more sense to have a selector similar to the fleet server hosts' one (in the add fleet server flyout):
![Screenshot 2024-06-20 at 16 02 15](https://github.com/elastic/kibana/assets/16084106/02127446-c921-44a7-8bd0-980048ada331)
Something like this would be much clearer for the user, as they could add as many existing policies AND a new agent policy at the same time.

criamico commented:
@elasticmachine merge upstream

kpollich commented on x-pack/plugins/fleet/common/services/agent_policies_helpers.ts:
```diff
@@ -44,3 +44,10 @@ function policyHasIntegration(agentPolicy: AgentPolicy, packageName: string) {
 
   return agentPolicy.package_policies?.some((p) => p.package?.name === packageName);
 }
+
+export function getInheritedNamespace(agentPolicies: AgentPolicy[]): string {
+  if (agentPolicies.length === 1) {
+    return agentPolicies[0].namespace;
+  }
+  return 'default';
+}
```
Could we write a quick unit test for this? Seems like a good candidate for simple tests.

criamico replied:
Added one unit test in https://github.com/elastic/kibana/pull/186084/commits/191ad44c2168f7c43352d664c1a44f221d095380

kpollich reviewed (approved):
Code + tests + screen recordings LGTM. Let's move discussion around the policy selection UX to another issue and maybe have Simona weigh in as I believe she designed the input this way intentionally. 

Added one note about an opportunity to add a quick unit test but otherwise test coverage LGTM. 🚀

criamico commented:
@kpollich I opened a new issue to capture our discussion around the policy selector: https://github.com/elastic/kibana/issues/186629
